core/src/data_sources/data_source.rs

```diff
@@ -594,37 +596,156 @@ impl DataSource {
             utils::now() - now
         ));
 
-        // Chunk splits into a vectors of 8 chunks (Vec<Vec<String>>)
-        let chunked_splits = splits
-            .chunks(8)
-            .map(|chunk| chunk.to_vec())
-            .collect::<Vec<_>>();
+        let current_doc = store
+            .load_data_source_document(
+                &self.project,
+                &self.data_source_id(),
+                &document_id.to_string(),
+                &None,
+            )
+            .await?;
 
+        let mut should_embbed = match current_doc {
+            Some(current_doc) => current_doc.hash != document_hash,
+            None => true,
+        };
         // Ordered results with offset, stirng and vector
         let mut e: Vec<(usize, String, EmbedderVector)> = vec![];
-
-        let now = utils::now();
         let mut offset: usize = 0;
-        // Embed batched chunks sequentially.
-        for chunk in chunked_splits {
-            let r = EmbedderRequest::new(
-                self.config.provider_id.clone(),
-                &self.config.model_id,
-                chunk.iter().map(|s| s.as_str()).collect::<Vec<_>>(),
-                self.config.extras.clone(),
-            );
+        if should_embbed == false {
+            let mut cached_embeddings: HashMap<String, EmbedderVector> = HashMap::new();
+            let mut page_offset: Option<PointId> = None;
+            loop {
```

Comment:
Can we add a safety net here (maybe if  we looped 512 time we exit) with a simple counter. I get real antsy when I see a `loop {}`